workshops/web_chat/README.md

```diff
@@ -0,0 +1,782 @@
+    try {
+        const { conn, r: bufReader, w: bufWriter, headers } = req
+        let socket = await acceptWebSocket({
+            conn,
+            bufReader,
+            bufWriter,
+            headers,
+        })
+    } catch (error) {
+        if (req.url === "/") {
+            req.url = "/index.html"
+        }
+        try {
+            const data = await Deno.readTextFile(req.url.substring(1))
+            await req.respond({ status: 200, body: data })
+        } catch {
+            await req.respond({ status: 404 })
+        }  
+    }
+}
+```
+
+Every time someone vists our page, we'll want to send index.html to them. However, trying to connect to a websocket also counts as a request. For every request that we get, we want to try to parse it as a websocket request. If that fails, then we send the index.html. If you hit "Run" now, you should see... no changes. This is because we haven't added any code to connect to our websocket in the frontend yet! Add the following code to frontend.js:
+
+```js
+let ws
+
+document.addEventListener("DOMContentLoaded", _ => {
+    ws = new WebSocket(`wss://${window.location.host}`)
+    ws.onopen = function (event) {
+        console.log("Socket connection is open")
+        ws.send(JSON.stringify({
+            type: "open",
+            data: {}
+        }))
+    }
+
+    ws.onmessage = function (event) {
+      console.log("Message received")
+      const msg = JSON.parse(event.data)
+      console.log(msg.data)
+    }
+})
+```
+
+If you try to run the code you should see... no changes (again 😑). The socket connects to the server but the server doesn't do anything. Add the following code just before `} catch (error) {` in index.ts:
+
+```js
+try {
+    handleWs(socket)
+} catch (err) {
+    console.error(`failed to receive frame: ${err}`)
+    if (!socket.isClosed) {
+        await socket.close(1000).catch(console.error)
+    }
+}
+```
+
+At the end of index.ts, add:
+
+```js
+async function handleWs(socket: WebSocket) {
+    for await (const event of socket) {
+        if (typeof event === "string") {
+            const parsedEvent = JSON.parse(event)
+            if (parsedEvent.type === "open") {
+                console.log("Connection established with a client.")
+            }
+        }
+    }
+}
+```
+
+Explanation:
+(from above) Every time someone connects to our websocket, we want to try to handle the websocket, i.e. we want to do something for every request to the websocket. If an error is present, we log the error and close the connection. Here, we're listening for an event of type "open" (which we're sending from index.html). If you try running your code now, you should see "Connection established with a client." in your console 🎉🎉
+
+Next, let's store connections to users and send a welcome message to them. We'll create a new array, i.e. a list (add below ``console.log(`Chat server is running on 8080`)``):
+
+```js
+let users: WebSocket[] = []
```

Comment:
Where should this go? Not 100% clear

Reply:
Kinda confused here too, it says "add below ``console.log(`Chat server is running on 8080`)``", what's not clear here?